docs/contributing/code-reviewing.md

```diff
@@ -1,231 +1,364 @@
+
+1. Make sure the PR **avoids duplicated code.** Code duplication is a huge
+   source of bugs in large projects and makes the codebase difficult to
+   understand, so we avoid significant code duplication wherever possible.
+   Sometimes avoiding code duplication involves some refactoring of existing
+   code; if so, that should usually be done as its own series of commits (not
+   squashed into other changes or left as a thing to do later). That series of
+   commits can be in the same pull request as the feature that they support, and
+   we recommend ordering the history of commits so that the refactoring comes
+   _before_ the feature. That way, it's easy to merge the refactoring (and
+   minimize risk of merge conflicts) if there are still user experience issues
+   under discussion for the feature itself.
+
+1. **Good comments** help. It's often worth thinking about whether explanation
+   in a commit message or pull request discussion should be included in
+   a comment, `/docs`, or other documentation. But it's better yet if
+   verbose explanation isn't needed. We prefer writing code that is
+   readable without explanation over a heavily commented codebase using
+   lots of clever tricks.
+
+1. Make sure the PR follows Zulip's **coding style**. See the Zulip [code-style]
+   documentation for details. Our goal is to have as much of this as possible
+   verified via the linters and tests, but there will always be unusual forms of
+   Python/JavaScript style that our tools don't check for.
+
+1. If you can, step back and think about the **technical design**. There are a
+   lot of considerations here: security, migration paths/backwards compatibility,
+   cost of new dependencies, interactions with features, speed of performance,
+   API changes. Security is especially important and worth thinking about
+   carefully with any changes to security-sensitive code like views.
+
+**Think about testing:**
+
+1. **The CI build tests need to pass.** One can investigate
+   any failures and figure out what to fix by clicking on a red X next
+   to the commit hash or the Detail links on a pull request. (Example:
+   in [#17584](https://github.com/zulip/zulip/pull/17584),
+   click the red X before `49b10a3` to see the build jobs
+   for that commit. You can see that there are 7 build jobs in total.
+   All the 7 jobs run in GitHub Actions. You can see what caused
+   the job to fail by clicking on the failed job. This will open
+   up a page in the CI that has more details on why the job failed.
+   For example [this](https://github.com/zulip/zulip/runs/2092955762)
+   is the page of the "Debian 10 Buster (Python 3.7, backend + frontend)" job.
+   See our docs on [continuous integration](../testing/continuous-integration.md)
+   to learn more.
+
+2. Make sure **the code is well-tested**; see below for details. The PR should
+   summarize any [manual testing](#manual-testing) that was done to validate
+   that the feature is working as expected.
```

Comment:
Maybe mention that before/after screencasts are a great way to make it easy to verify manual testing that was done.

Reply:
I tried, but decided that it doesn't belong here.